src/java.base/share/classes/java/util/AbstractMap.java

```diff
@@ -407,46 +372,41 @@ public boolean contains(Object k) {
      * method will not all return the same collection.
      */
     public Collection<V> values() {
-        Collection<V> vals = values;
-        if (vals == null) {
-            vals = new AbstractCollection<V>() {
-                public Iterator<V> iterator() {
-                    return new Iterator<V>() {
-                        private Iterator<Entry<K,V>> i = entrySet().iterator();
-
-                        public boolean hasNext() {
-                            return i.hasNext();
-                        }
-
-                        public V next() {
-                            return i.next().getValue();
-                        }
-
-                        public void remove() {
-                            i.remove();
-                        }
-                    };
-                }
+        return new AbstractCollection<>() {
```

Comment:
Note that this causes `m.values().equals(m.values())` to no longer return `true`, as base `Collection::equals` is defined using reference equality.

Reply:
Looking at the specification above: 
> No synchronization is performed, so there is a slight chance that multiple calls to this method will not all return the same collection.

This should be an acceptable change to behavior.

Reply:
This is brought to attention in the CSR.